core/protocols/smithy-rpcv2-protocol/src/main/java/software/amazon/awssdk/protocols/rpcv2/internal/SdkRpcV2CborGenerator.java

```diff
@@ -0,0 +1,120 @@
+import software.amazon.awssdk.protocols.json.StructuredJsonGenerator;
+import software.amazon.awssdk.thirdparty.jackson.core.JsonFactory;
+import software.amazon.awssdk.thirdparty.jackson.dataformat.cbor.CBORGenerator;
+
+/**
+ * Thin wrapper around Jackson's JSON generator for CBOR.
+ */
+@SdkInternalApi
+public final class SdkRpcV2CborGenerator extends SdkJsonGenerator {
+
+    private static final int CBOR_TAG_TIMESTAMP = 1;
+    private static final int MILLI_SECOND_PRECISION = 3;
+
+    SdkRpcV2CborGenerator(JsonFactory factory, String contentType) {
+        super(factory, contentType);
+    }
+
+    /**
+     * Jackson doesn't have native support for timestamp. As per the RFC 7049 (https://tools.ietf.org/html/rfc7049#section-2.4.1)
+     * we will need to write a tag and write the epoch.
+     */
+    @Override
+    public StructuredJsonGenerator writeValue(Instant instant) {
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeTag(CBOR_TAG_TIMESTAMP);
+            BigDecimal dateValue = BigDecimal.valueOf(instant.toEpochMilli());
+            generator.writeNumber(dateValue.scaleByPowerOfTen(0 - MILLI_SECOND_PRECISION).doubleValue());
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    public StructuredJsonGenerator writeStartArray(int size) {
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeStartArray(null, size);
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    public StructuredJsonGenerator writeValue(double val) {
+        if (canConvertToLong(val)) {
+            return writeValue((long) val);
+        }
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeNumber(val);
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    public StructuredJsonGenerator writeValue(float val) {
+        if (canConvertToLong(val)) {
+            return writeValue((long) val);
+        }
+        CBORGenerator generator = getGenerator();
+        try {
+            generator.writeNumber(val);
+        } catch (IOException e) {
+            throw new JsonGenerationException(e);
+        }
+        return this;
+    }
+
+    @Override
+    protected CBORGenerator getGenerator() {
+        return (CBORGenerator) super.getGenerator();
+    }
+
+    /**
+     * Checks if we can convert the floating point value to a long. If we can we then use the writeNumber(long), alongside with
+     * having enabled Feature.WRITE_MINIMAL_INTS will allow us to represent the floating point values minimally.
+     */
+    private static boolean canConvertToLong(double value) {
+        return value % 1 == 0
```

Comment:
(I think (double) longValue != value would be faster than value % 1 on a sufficiently advanced CPU. Worth benchmarking?)